include/rmm/mr/device/detail/arena.hpp

```diff
@@ -33,24 +33,48 @@ namespace mr {
 namespace detail {
 namespace arena {
 
-constexpr std::size_t superblock_size = 1u << 26u;  ///< Size of a superblock (64 MiB).
+/// Minimum size of a superblock (32 KiB).
+constexpr std::size_t minimum_superblock_size = 1u << 15u;
 
 /**
  * @brief Represents a chunk of memory that can be allocated and deallocated.
  *
- * A fixed-sized block obtained from the global arena is called a "superblock". Only a superblock
- * can be returned to the global arena.
+ * A block bigger than a certain size is called a "superblock".
  */
-struct block {
-  void* pointer{};     ///< Raw memory pointer.
-  std::size_t size{};  ///< Size in bytes.
-  bool is_head{};      ///< Indicates whether pointer was allocated from upstream.
+class block {
+ public:
+  /**
+   * @brief Construct a default block.
+   */
+  block() = default;
+
+  /**
+   * @brief Construct a block given a pointer and size.
+   *
+   * @param pointer The address for the beginning of the block.
+   * @param size The size of the block.
+   */
+  block(char* pointer, size_t size) : pointer_(pointer), size_(size) {}
+
+  /**
+   * @brief Construct a block given a void pointer and size.
+   *
+   * @param pointer The address for the beginning of the block.
+   * @param size The size of the block.
+   */
+  block(void* pointer, size_t size) : pointer_(static_cast<char*>(pointer)), size_(size) {}
+
+  /// Returns the underlying pointer.
+  void* pointer() const { return pointer_; }
+
+  /// Returns the size of the block.
+  size_t size() const { return size_; }
 
   /// Returns true if this block is valid (non-null), false otherwise.
-  bool is_valid() const { return pointer != nullptr; }
+  bool is_valid() const { return pointer_ != nullptr; }
 
   /// Returns true if this block is a superblock, false otherwise.
-  bool is_superblock() const { return is_head && size == superblock_size; }
+  bool is_superblock() const { return size_ >= minimum_superblock_size; }
```

Comment:
Is this correct? If you had a superblock allocation of 1MiB, and then that block is split in half, then both would be considered superblocks even though they are part of the same upstream allocation. Deallocating the second half could fail.

Reply:
With this PR free superblocks are not kept in the per-thread arena but are always returned to the global arena, so if an allocation is bigger than the superblock threshold it's always returned when freed.

Reply:
Where does this happen?

Reply:
In `shrink_arena()`.